show latest major version available in `gleam deps update`

andho opened the pull request:
This PR aims to solve #3843.
The sample output from that issue:
```bash
 % gleam deps update
  Resolving versions
Downloading packages
 Downloaded 2 packages in 0.01s
 
Hint: the following dependencies have new major versions available...

- wibble@2
- wobble@3
```

# Todos
- [x] ~~Handle Async: fireoff the checks and show the results at the end.~~ Didn't make sense to do this.
- [ ] Telemetry? Not sure it's needed for this optional part.
- [x] ~~Merge version check into a single call~~ Hexpm API doesn't have this. Went with cache instead.
- [x] Get the direct dependencies only

andho commented:
> Hello! Are you still working on this one? Thanks

I was hoping to work on it this weekend but some things have come up. I can only get back to it next weekend.

andho commented:
Just the major version hist without any other output is weird though:
![image](https://github.com/user-attachments/assets/af1b4d01-5a90-4782-9016-7cdb6a2707fd)

Helpful, but looks weird.

inoas commented:
Could a flag be provided that forces a non-zero exit-code if major deps are not bumped?

lpil commented on compiler-core/src/dependency.rs:
```diff
@@ -832,4 +970,91 @@ mod tests {
+            ]
+            .into_iter()
+            .collect(),
+            //requirements: HashMap::new(),
```
Unused code here

andho replied:
Haha, this is embarrassing.

lpil commented on compiler-cli/src/dependencies.rs:
```diff
@@ -981,6 +1014,15 @@ impl dependency::PackageFetcher for PackageFetcher {
+            let result = runtime_cache.get(package);
+
+            if let Some(result) = result {
+                return Ok(result.clone());
```
Rather than clone the entire package lets record the highest version here.

andho replied:
Oh! This function is originally used by the `DependencyProvider`. But as the major version need the same info, but to avoid multiple http calls to hexpm, I've added caching here.

I think the way to avoid cloning would be to have PackageFetcher own package (and return a ref) but that would have to touch the dependency resolution part so I didn't want to include in this PR (increased scope).

andho replied:
Here's an implementation with PackageFetcher owning packages and RefCounting: https://github.com/andho/gleam/commit/851b9a6b309e9739f1d2b44e8178831f378f00eb

andho replied:
Ah I just realized I never wired up the same `PackageFetcher` to be used in other parts of the dependency resolution so the cache is not even used. I'll see about getting that wired up soon.

andho replied:
`PackageFetcher` is now passed by reference in all instances and instantiated in `list` and `download` only.

lpil replied:
It would be best to do that as to not clone everything. Do you think it would be complex or involved?

andho replied:
Changed to pass package as refs using Rc. DependencyProvider still needs to clone the package because it sorts the releases field before passing to pubgrub.

lpil commented on compiler-core/src/dependency.rs:
```diff
@@ -63,6 +64,65 @@ where
+            let Some(latest) = &hexpackage
+                .releases
+                .iter()
+                .map(|release| release.version.clone())
```
Can take a reference here and avoid cloning

andho replied:
Fixed

lpil commented on compiler-core/src/dependency.rs:
```diff
@@ -63,6 +64,65 @@ where
+                .iter()
+                .map(|release| release.version.clone())
+                .filter(|version| !version.is_pre())
+                .sorted_by(|a, b| b.cmp(a))
+                .next()
```
Using `.max` will be faster than sorting and taking the first

andho replied:
Fixed.

lpil commented on compiler-cli/src/dependencies.rs:
```diff
@@ -285,9 +286,31 @@ pub fn download<Telem: Telemetry>(
+
+        // lazily assuming the longest version could be 8 characters. eg: 1.1234.2
+        // excluding qualifiers other than x.y.z
+        eprintln!("{name}:{padding} {v1:<8} -> {v2:<8}");
```
Let's render this to a string and then snapshot test the function. This function can then just print the string and not be responsible for formatting and printing

andho replied:
Done.

lpil reviewed (commented):
Very nice! I've left some small notes inline. Could you update the changelog too please 🙏

andho commented:
> Could a flag be provided that forces a non-zero exit-code if major deps updates are available?
> 
> Would that make sense?
> 
> The idea would be to have a CI check that goes red if deps are not up to date.

@inoas  I think the flag would make more sense in a `gleam deps outdated` command instead.

lpil reviewed (commented):
Great, thank you.

<img width="660" alt="image" src="https://github.com/user-attachments/assets/b3ea1b12-8390-4edb-84fb-7e61b35aa642" />

I gave it a try and while the information is there and useful I don't think it fits in at all with any of the other information printed. How can we show this better?

Should it be shown for all resolution or only for `gleam update`? If the latter it would be easier to make a good output format for it.

andho commented:
> Should it be shown for all resolution or only for `gleam update`?

It makes sense to show for `gleam update` and `gleam deps download`

Although it was helpful when I recently did `gleam build` and I saw that `decode` package was updated to 1.0.0. But due to the execution order it would take some rearranging to get this output to come last.

I'd say for an MVP show only in `gleam update` and `gleam deps download`. I'll only have to add a flag to `download` fn.

lpil commented:
Sounds great!

lpil commented on compiler-cli/src/dependencies.rs:
```diff
@@ -219,73 +221,75 @@ pub fn download<Telem: Telemetry>(
     // packages being resolved (not the locked ones).
     use_manifest: UseManifest,
+    // If true we check for major version updates and print them to the console.
+    check_major_versions: bool,
```
Create an enum instead of using bools please. `true` and `false` mean nothing without context so they are hard to understand. See the other arguments for examples.

lpil commented on compiler-cli/src/dependencies.rs:
```diff
@@ -219,73 +221,75 @@ pub fn download<Telem: Telemetry>(
-        tracing::debug!("writing_manifest_toml");
-        write_manifest_to_disc(paths, &manifest)?;
+    output_string
+        .push_str("\nHint: the following dependencies have new major versions available:\n\n");
```
```suggestion
        .push_str("\nThe following dependencies have new major versions available:\n\n");
```

lpil commented on compiler-core/src/dependency.rs:
```diff
@@ -63,6 +63,61 @@ where
+/**
+* Used to compare 2 versions of a package.
+*/
+pub type PackageVersionDiffs = HashMap<String, (Version, Version)>;
```
Let's either use this always or remove it

andho replied:
> Let's either use this always or remove it

I'm not sure what you mean. `PackageVersionDiffs` are used only for the major versions check currently. I don't see any other place making comparisons between 2 versions (except probably within pubgrub which makes comparisons betweens many versions probably).

Is you suggestion to remove the type and use `HashMap<String, (Version, Version)>` everywhere?

lpil replied:
Either always use it or always using the underlying type is fine. We don't want a mix.

lpil commented on compiler-cli/src/dependencies.rs:
```diff
@@ -543,6 +491,237 @@ fn same_requirements(
+    fn should_use_manifest(mut self) -> Self {
+        self.use_manifest = UseManifest::Yes;
+        self
+    }
```
This is cool!

Another option which can be convenient at times is to have a struct with public fields and have an `into_dependency_manager()` method or something like that

andho replied:
I looked into builder pattern and init pattern (misnamed but similar to what you're suggesting). But as the tokio runtime, Package Fetcher and Telemetry needs to be specified, I cannot implement Default. Which means I'll have to define all struct fields at callsite. Which means any addition or change to the struct will require changes at all callsites.

Alternatively I could go for a builder, but seems too small right now.

andho replied:
Although...I could make it `DependencyManagerConfig.into_dependecy_manager(runtime: ..., package_fetcher: ..., telemetry: ...)`.

lpil reviewed (commented):
Looks great!! Thank you. I've left two tiny notes inline.

lpil commented:
Are you still planning to work on this? Thank you 🙏

andho commented:
> Are you still planning to work on this? Thank you 🙏

Hey yeah, will be working on this 🙂. Was a bit busy over the last 2 weeks, will be resuming it this week 🤞

lpil commented on compiler-cli/src/build.rs:
```diff
@@ -10,14 +10,19 @@ use gleam_core::{
+        telemetry,
+        None,
+        Vec::new(),
+        dependencies::DependencyManagerConfig::default(),
```
This is much less clear than before- please explicitly state each item of config. The programmer should be able to tell what the desired behaviour is at each call-site. This will mean removing the default implementation.

andho replied:
Done

lpil commented on compiler-cli/src/dependencies.rs:
```diff
@@ -543,6 +490,242 @@ fn same_requirements(
 }
 
+pub struct DependencyManagerConfig {
+    pub mode: Mode,
+    pub use_manifest: UseManifest,
+    pub check_major_versions: CheckMajorVersions,
```
Document what each of these do please 🙏

andho replied:
Done

lpil commented on compiler-cli/src/dependencies.rs:
```diff
@@ -543,6 +490,242 @@ fn same_requirements(
+    }
+}
+
+pub struct DependencyManager<Telem: Telemetry, P: dependency::PackageFetcher> {
```
Remove these constraints please 🙏 Constraints are only needed on the functions.

andho replied:
Done

lpil reviewed (commented):
Thank you! A few small things

lpil commented on compiler-cli/src/dependencies.rs:
```diff
@@ -678,6 +630,239 @@ fn same_requirements(
+pub struct DependencyManagerConfig {
+    /// In `Prod` mode, dev dependencies are not considered during the executed operation.
+    /// Otherwise (`Dev` or `Lsp`), all dependencies are considered
+    pub mode: Mode,
```
This has moved this internal mode value into the public interface now. What was your motivation for this? From what I can see it is never set to anything but `Dev`.

andho replied:
Yeah you're right. I didn't intend to move it into the public interface, but to bundle configuration related data into a single struct, to reduce number of arguments passed around.

For now I'll remove `mode` from `DependencyManagerConfig`.

andho replied:
I have also made `DependencyManager` private as it's not directly used outside of `dependencies`.

lpil commented on compiler-cli/src/dependencies.rs:
```diff
@@ -678,6 +633,237 @@ fn same_requirements(
+        }
+    }
+
+    fn download(
```
It was tricky for me to determine that this was the public API of this class. Could you mark this as public and move the class to sub-module please 🙏 

I'm concerned about the API being understandable, particularly since the locking has been moved inside this class, so it is extremely important that the correct entrypoint is used.

andho replied:
Done.

lpil reviewed (commented):
Looking great! One more thing

lpil reviewed (commented):
Beautiful work, thank you so much!

I've been manually testing this and it's really nice and useful, but it prints about dependencies being out of date before they are reported as having been added.

```
$ gleam add lustre@2 gleam_json@1
  Resolving versions
Downloading packages
 Downloaded 5 packages in 0.48s

Hint: the following dependencies have new major versions available:

gleam_json 1.0.1 -> 2.3.0
lustre     2.0.1 -> 4.6.3

      Added lustre v2.0.1
      Added gleam_json v1.0.1
```

I think it would make more sense the other way around

```
$ gleam add lustre@2 gleam_json@1
  Resolving versions
Downloading packages
 Downloaded 5 packages in 0.48s
      Added lustre v2.0.1
      Added gleam_json v1.0.1

The following dependencies have new major versions available:

gleam_json 1.0.1 -> 2.3.0
lustre     2.0.1 -> 4.6.3
```

What do you think?

andho commented:
> Beautiful work, thank you so much!
> 
> I've been manually testing this and it's really nice and useful, but it prints about dependencies being out of date before they are reported as having been added.
> 
> ```
> $ gleam add lustre@2 gleam_json@1
>   Resolving versions
> Downloading packages
>  Downloaded 5 packages in 0.48s
> 
> Hint: the following dependencies have new major versions available:
> 
> gleam_json 1.0.1 -> 2.3.0
> lustre     2.0.1 -> 4.6.3
> 
>       Added lustre v2.0.1
>       Added gleam_json v1.0.1
> ```
> 
> I think it would make more sense the other way around
> 
> ```
> $ gleam add lustre@2 gleam_json@1
>   Resolving versions
> Downloading packages
>  Downloaded 5 packages in 0.48s
>       Added lustre v2.0.1
>       Added gleam_json v1.0.1
> 
> The following dependencies have new major versions available:
> 
> gleam_json 1.0.1 -> 2.3.0
> lustre     2.0.1 -> 4.6.3
> ```
> 
> What do you think?

Yeah the current output is confusing. But that's actually a bug. Major version availability should be shown only on `update` and `deps download` command as we discussed. I'll take a look.

PS. I'm traveling until 28th feb, so updates might be a bit slow.

andho commented:
Hi @lpil. I'm not able to reproduce the issue. I think you might have run an old version. "Hint:" had been removed from the output a few commits back.

lpil commented:
Sorry I'm not sure what issue you mean 😅

andho commented:
Haha, yeah it's been a while, sorry!

I'm referring to this:
> I've been manually testing this and it's really nice and useful, but it prints about dependencies being out of date before they are reported as having been added.
> 
> ```
> $ gleam add lustre@2 gleam_json@1
>   Resolving versions
> Downloading packages
>  Downloaded 5 packages in 0.48s
> 
> Hint: the following dependencies have new major versions available:
> 
> gleam_json 1.0.1 -> 2.3.0
> lustre     2.0.1 -> 4.6.3
> 
>       Added lustre v2.0.1
>       Added gleam_json v1.0.1
> ```
> 
> I think it would make more sense the other way around
> 
> ```
> $ gleam add lustre@2 gleam_json@1
>   Resolving versions
> Downloading packages
>  Downloaded 5 packages in 0.48s
>       Added lustre v2.0.1
>       Added gleam_json v1.0.1
> 
> The following dependencies have new major versions available:
> 
> gleam_json 1.0.1 -> 2.3.0
> lustre     2.0.1 -> 4.6.3
> ```
> 
> What do you think?

In the latest iteration, the major version check is only triggered for `update` and `deps download` commands. So it won't show up in `add` command.

lpil reviewed (commented):
Hello! Sorry for taking so long to get back to you.

I tried to rebase this on main but I got confused as I wasn't familiar enough with the code. Could you rebase this up to date for me please 🙏 Both squashing or preserving history is fine

andho commented:
@lpil I've rebased on main. Here's a summary of the changes:

1. Added `CheckMajorVersions` enum as a flag to indicate if gleam should check for major version updates.
2. Changed `PackageFetcher` in dependencies to not be boxed so that it's monomorphized instead of using dynamic dispatch.
3. Changed `PackageFetcher` to cache the package info retrieved from hex, so that duplicate calls to hexpm won't be made (during version resolution and then during major version check).
4. Addition of `CheckMajorVersions` caused `too many parameters` lint issue, so those **configuration/context** related values are now moved to `DependencyManager` struct, and the 3 functions that had `too many parameters` has been moved into an impl for `DependencyManager`. Those functions are `download`, `get_manifest` and `resolve_versions`.
   1. I have manually compared the current `main` branch and this PR branch, and there are no difference in those functions between the 2. Manual check is necessary as the methods were moved out to `compiler-cli/src/dependencies/dependency_manager.rs`.
5. `CheckMajorVersions` is enabled only for `gleam [deps] update` and `gleam download`.

lpil reviewed (approved):
Thank you so much!!!! Sorry for taking to long to merge this
